resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala

```diff
@@ -320,50 +322,83 @@ private[spark] class KubernetesClusterSchedulerBackend(
-          if (podIP != null) {
-            executorPodsByIPs.remove(podIP)
+          failedInitExecutors.add(executorId)
+          if (failedInitExecutors.size >= executorMaxInitErrors) {
+            val errorMessage = s"Aborting Spark application because $executorMaxInitErrors" +
+              s" executors failed to start. The maximum number of allowed startup failures is" +
+              s" $executorMaxInitErrors. Please contact your cluster administrator or increase" +
+              s" your setting of ${KUBERNETES_EXECUTOR_MAX_INIT_ERRORS.key}."
+            logError(errorMessage)
+            KubernetesClusterSchedulerBackend.this.scheduler.sc.stopInNewThread()
+            throw new SparkException(errorMessage)
           }
+          handleFailedPod(action, pod, podName, podIP)
 
-          val executorExitReason = if (action == Action.ERROR) {
-            logWarning(s"Received error event of executor pod $podName. Reason: " +
-              pod.getStatus.getReason)
-            executorExitReasonOnError(pod)
-          } else if (action == Action.DELETED) {
-            logWarning(s"Received delete event of executor pod $podName. Reason: " +
-              pod.getStatus.getReason)
-            executorExitReasonOnDelete(pod)
-          } else {
-            throw new IllegalStateException(
-              s"Unknown action that should only be DELETED or ERROR: $action")
-          }
-          podsWithKnownExitReasons.put(pod.getMetadata.getName, executorExitReason)
-
-          if (!disconnectedPodsByExecutorIdPendingRemoval.containsKey(executorId)) {
-            log.warn(s"Executor with id $executorId was not marked as disconnected, but the " +
-              s"watch received an event of type $action for this executor. The executor may " +
-              "have failed to start in the first place and never registered with the driver.")
-          }
-          disconnectedPodsByExecutorIdPendingRemoval.put(executorId, pod)
+        case Action.DELETED | Action.ERROR =>
+          handleFailedPod(action, pod, podName, podIP)
 
         case _ => logDebug(s"Received event of executor pod $podName: " + action)
       }
     }
 
+    private def handleFailedPod(action: Action, pod: Pod, podName: String, podIP: String) = {
+      val executorId = getExecutorId(pod)
+      logDebug(s"Executor pod $podName at IP $podIP was at $action.")
+      if (podIP != null) {
+        executorPodsByIPs.remove(podIP)
+      }
+
+      val executorExitReason = if (action == Action.ERROR) {
+        logWarning(s"Received error event of executor pod $podName. Reason: " +
+          pod.getStatus.getReason)
+        executorExitReasonOnError(pod)
+      } else if (action == Action.DELETED) {
+        logWarning(s"Received delete event of executor pod $podName. Reason: " +
+          pod.getStatus.getReason)
+        executorExitReasonOnDelete(pod)
+      } else if (action == Action.MODIFIED) {
+        executorExitReasonOnInitError(pod)
```

Comment:
I think `Action.MODIFIED` can be fired for a lot of other reasons. I was thinking that we should just use container exit status, and have that reflect either the main/init container status depending on which container failed.

Reply:
I would suggest separating out handling of pod that has failed init-container(s) into its own method.

Reply:
What's the full set of error types that we should be tracking here? Are there docs that have a full enumerated list?

Reply:
I think this https://kubernetes.io/docs/tasks/debug-application-cluster/debug-init-containers/#understanding-pod-status.

Reply:
That only covers init-containers - I think @foxish is referring to other classes of initialization errors, like failing to pull the image.

Reply:
There isn't a doc, but I'm putting together an initial list. We can grow it as we discover more during operations. Would be good to 0/1 nodes are available: 1 node(s) had disk pressure.